include/envoy/stats/symbol_table.h

```diff
@@ -1,13 +1,163 @@
+   * allocate a StatName with the correct amount of storage.
+   *
+   * When a name is encoded, it bumps reference counts held in the table for
+   * each symbol. The caller is responsible for creating a StatName using this
+   * SymbolEncoding and ultimately disposing of it by calling
+   * SymbolTable::free(). Users are protected from leaking symbols into the pool
+   * by ASSERTions in the SymbolTable destructor.
+   *
+   * @param name The name to encode.
+   * @return SymbolEncoding the encoded symbols.
+   */
+  virtual SymbolEncoding encode(absl::string_view name) PURE;
+
+  /**
+   * @return uint64_t the number of symbols in the symbol table.
+   */
+  virtual uint64_t numSymbols() const PURE;
+
+  /**
+   * Decodes a vector of symbols back into its period-delimited stat name. If
+   * decoding fails on any part of the symbol_vec, we release_assert and crash,
+   * since this should never happen, and we don't want to continue running
+   * with a corrupt stats set.
+   *
+   * @param stat_name the stat name.
+   * @return std::string stringifiied stat_name.
+   */
+  virtual std::string toString(const StatName& stat_name) const PURE;
+
+  /**
+   * Deterines whether one StatName lexically precedes another. Note that
+   * the lexical order may not exactly match the lexical order of the
+   * elaborated strings. For example, stat-name of "-.-" would lexically
+   * sort after "---" but when encoded as a StatName would come lexically
+   * earlier. In practice this is unlikely to matter as those are not
+   * reasonable names for Envoy stats.
+   *
+   * Note that this operation has to be performed with the context of the
+   * SymbolTable so that the individual Symbol objects can be converted
+   * into strings for lexical comparison.
+   *
+   * @param a the first stat name
+   * @param b the second stat name
+   * @return bool true if a lexically precedes b.
+   */
+  virtual bool lessThan(const StatName& a, const StatName& b) const PURE;
+
+  /**
+   * Joins two or more StatNames. For example if we have StatNames for {"a.b",
+   * "c.d", "e.f"} then the joined stat-name matches "a.b.c.d.e.f". The
+   * advantage of using this representation is that it avoids having to
+   * decode/encode into the elaborated form, and does not require locking the
+   * SymbolTable.
+   *
+   * The caveat is that this representation does not bump reference counts on
+   * the referenced Symbols in the SymbolTable, so it's only valid as long for
+   * the lifetime of the joined StatNames.
+   *
+   * This is intended for use doing cached name lookups of scoped stats, where
+   * the scope prefix and the names to combine it with are already in StatName
+   * form. Using this class, they can be combined without acessingm the
+   * SymbolTable or, in particular, taking its lock.
+   */
+  virtual StoragePtr join(const StatName& a, const StatName& b) const PURE;
+  virtual StoragePtr join(const std::vector<StatName>& stat_names) const PURE;
```

Comment:
Probably premature optimization, but the vector one requires allocating a temp, and I was thinking the 2-arg case was going to be common enough to justify the extra end-point.

Reply:
done; for the record I doubt we'd find this in a profile, but it could be exposed in a microbenchmark. It's possible that std::vector has optimized the 2-element case to avoid allocations. This will be in the hot-path.